pkg/controller/quota/quota.go

```diff
@@ -0,0 +1,168 @@
+		status.Unknown("waiting for quota allocation")
+	} else if quotaRequest.Status.Condition(adminv1.QuotaRequestCondition).Success {
+		status.Success()
+	}
+
+	return nil
+}
+
+// EnsureQuotaRequest ensures that the quota request exists and is up to date.
+func EnsureQuotaRequest(req router.Request, resp router.Response) error {
+	appInstance := req.Object.(*apiv1.AppInstance)
+
+	// Don't do anything if quota isn't enabled for this project
+	if enforced, err := isEnforced(req, appInstance.Namespace); !enforced {
+		return err
+	}
+
+	// Create the quota request object and give calculate the standard numeric values
+	name, namespace, app := appInstance.Name, appInstance.Namespace, appInstance.Status.AppSpec
+	quotaRequest := &adminv1.QuotaRequestInstance{
+		ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: namespace},
+		Spec: adminv1.QuotaRequestInstanceSpec{
+			Resources: adminv1.Resources{
+				Containers: len(app.Containers),
+				Jobs:       len(app.Jobs),
+				Volumes:    len(app.Volumes),
+				Secrets:    len(app.Secrets),
+				Images:     len(app.Images),
+			},
+		},
+	}
+
+	status := condition.Setter(appInstance, resp, apiv1.AppInstanceConditionQuotaAllocated)
+
+	// Add the more complex values to the quota request
+	addCompute(app.Containers, appInstance, quotaRequest)
+	addCompute(app.Jobs, appInstance, quotaRequest)
+	if err := addStorage(appInstance, quotaRequest); err != nil {
+		status.Error(err)
+		return err
+	}
+
+	resp.Objects(quotaRequest)
+	return nil
+}
+
+// addCompute adds the compute resources of the containers passed to the quota request.
+func addCompute(containers map[string]apiv1.Container, appInstance *apiv1.AppInstance, quotaRequest *adminv1.QuotaRequestInstance) {
+	// For each workload, add their memory/cpu requests to the quota request
+	for name, container := range containers {
+		var requirements corev1.ResourceRequirements
+		if specific, ok := appInstance.Status.Scheduling[name]; ok {
+			requirements = specific.Requirements
+		} else if all, ok := appInstance.Status.Scheduling[""]; ok {
+			requirements = all.Requirements
+		}
+
+		quotaRequest.Spec.Resources.CPU.Add(requirements.Requests["cpu"])
+		quotaRequest.Spec.Resources.Memory.Add(requirements.Requests["memory"])
+
+		// Recurse over any sidecars. Since sidecars can't have sidecars, this is safe.
+		addCompute(container.Sidecars, appInstance, quotaRequest)
+	}
+}
+
+// addStorage adds the storage resources of the volumes passed to the quota request.
+func addStorage(appInstance *apiv1.AppInstance, quotaRequest *adminv1.QuotaRequestInstance) error {
+	app := appInstance.Status.AppSpec
+
+	// Add the volume storage needed to the quota request. We only parse net new volumes, not
+	// existing ones that are then bound client-side.
+	for name, volume := range app.Volumes {
+		size := volume.Size
+		if bound, boundSize := boundVolumeSize(name, appInstance.Spec.Volumes); bound {
```

Comment:
If the volume is being bound, then it already exists. Would considering it here mean that the volume size is getting counted twice?

Reply:
Ah, I think that the name I made is a bit confusing. This logic isn't actually determining if we'll bind to any existing volumes, it just determines if any Acornfile defined Volumes will get overwritten sizes. That brings me to two action items:

1. Rename this function
2. Do some querying logic to determine if this will bind to some existing Volume.

Reply:
I don't think any querying logic is necessary: `appInstance.Spec.Volumes` will tell you if the volume is being bound to an existing volume (i.e. the `Volume` field will be set).

Reply:
Talked offline and this function is now updated to work properly. I've also added a new function to handle Secrets and their bindings in a similar way.